agent/functional_tests/tests/functionaltests_test.go

```diff
@@ -94,6 +94,42 @@ func TestOOMContainer(t *testing.T) {
 	}
 }
 
+// This test addresses a deadlock issue which was noted in GH:313 and fixed
+// in PR:320. It runs a service with 10 containers, waits for cleanup, starts
+// another two instances of that service and ensures that those tasks complete.
+func TestTaskCleanupDoesNotDeadlock(t *testing.T) {
+	// Set the ECS_ENGINE_TASK_CLEANUP_WAIT_DURATION to it's lowest permissible value
+	os.Setenv("ECS_ENGINE_TASK_CLEANUP_WAIT_DURATION", "60s")
+	defer os.Unsetenv("ECS_ENGINE_TASK_CLEANUP_WAIT_DURATION")
+
+	agent := RunAgent(t, nil)
+	defer agent.Cleanup()
+
+	// This bug was fixed in v1.8.1
+	agent.RequireVersion(">=1.8.1")
+
+	// Run two Tasks after cleanup, as the deadlock does not consistently occur after
+	// after just one task cleanup cycle.
+	for i := 0; i < 3; i++ {
+
+		// Start a task with ten containers
+		testTask, err := agent.StartTask(t, "ten-containers")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// 2 minutes should be enough for the Task to have completed. If the task has not
+		// completed and is in PENDING, the agent is most likely deadlocked.
+		err = testTask.WaitStopped(2 * time.Minute)
+		if err != nil {
+			t.Fatal("Task did not transition into to STOPPED in time.", err)
+		}
+
+		// Wait for the tasks to be cleaned up
+		time.Sleep(75 * time.Second)
```

Comment:
It would be good to either add an assertion after this sleep (maybe check the introspection API to see that the task is gone) or have the sleep be conditional for only the first two iterations (if you have no assertion, you're sleeping for no reason on the last iteration).